File: SeatingPlaceSuggestions/Java/SeatingPlaceSuggestions/src/main/java/org/weaveit/seatssuggestions/Row.java
package org.weaveit.seatssuggestions;

import java.util.ArrayList;
import java.util.List;

public record Row(String name, List<SeatingPlace> seatingPlaces) {

    public Row addSeatingPlace(SeatingPlace seatingPlace) {
        List<SeatingPlace> newSeatingPlaces = new ArrayList<>(seatingPlaces);
        newSeatingPlaces.add(seatingPlace);
        return new Row(this.name, newSeatingPlaces);
    }

    public SeatingOptionSuggested suggestSeatingOption(int partyRequested, PricingCategory pricingCategory) {
        for (SeatingPlace seat : seatingPlaces) {
            if (seat.isAvailable() && seat.matchCategory(pricingCategory)) {
                SeatingOptionSuggested seatingOptionSuggested = new SeatingOptionSuggested(partyRequested, pricingCategory);
                seatingOptionSuggested.addSeat(seat);

                if (seatingOptionSuggested.matchExpectation()) {
                    return seatingOptionSuggested;
                }
            }
        }

        return new SeatingOptionNotAvailable(partyRequested, pricingCategory);
    }

    public Row allocate(SeatingPlace seatingPlace) {
        List<SeatingPlace> newVersionOfSeats = new ArrayList<>();

        for (SeatingPlace currentSeat : seatingPlaces) {
            if (currentSeat.sameSeatingPlace(seatingPlace)) {
                newVersionOfSeats.add(new SeatingPlace(seatingPlace.rowName(), seatingPlace.number(), seatingPlace.pricingCategory(), SeatingPlaceAvailability.Allocated));
            } else {
                newVersionOfSeats.add(currentSeat);
            }
        }

        return new Row(seatingPlace.rowName(), newVersionOfSeats);
    }

}

File: SeatingPlaceSuggestions/CSharp/SeatingPlaceRecommender/Row.cs
using System.Collections.Generic;
using System.Linq;
using Value;

namespace SeatsSuggestions;

public class Row(string name, List<SeatingPlace> seatingPlaces) : ValueType<Row>
{
    public string Name { get; } = name;
    public List<SeatingPlace> SeatingPlaces { get; } = seatingPlaces;

    public Row AddSeatingPlace(SeatingPlace seatingPlace)
    {
        return new Row(Name, new List<SeatingPlace>(SeatingPlaces) { seatingPlace });
    }

    public SeatingOptionSuggested SuggestSeatingOption(SuggestionRequest suggestionRequest)
    {
        var seatingOptionSuggested = new SeatingOptionSuggested(suggestionRequest);

        foreach (var seatingPlace in SelectAvailableSeatingPlacesCompliantWith(suggestionRequest.PricingCategory))
        {
            seatingOptionSuggested.AddSeatingPlace(seatingPlace);

            if (seatingOptionSuggested.MatchExpectation()) return seatingOptionSuggested;
        }

        return new SeatingOptionNotAvailable(suggestionRequest);
    }

    private IEnumerable<SeatingPlace> SelectAvailableSeatingPlacesCompliantWith(PricingCategory pricingCategory)
    {
        return SeatingPlaces.Where(seatingPlace => seatingPlace.IsAvailable() && seatingPlace.MatchCategory(pricingCategory));
    }

    public Row Allocate(SeatingPlace seatingPlace)
    {
        var newVersionOfSeats = new List<SeatingPlace>();

        foreach (var currentSeat in SeatingPlaces)
            if (currentSeat.SameSeatingPlace(seatingPlace))
                newVersionOfSeats.Add(new SeatingPlace(seatingPlace.RowName, seatingPlace.Number, seatingPlace.PricingCategory,
                    SeatingPlaceAvailability.Allocated));
            else
                newVersionOfSeats.Add(currentSeat);

        return new Row(seatingPlace.RowName, newVersionOfSeats);
    }

    protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
    {
        return new object[] { Name, new ListByValue<SeatingPlace>(SeatingPlaces) };
    }
}